File: scripts/import-to-supabase.mjs
/**
 * JSON → Supabase (service role / secret key)
 * node scripts/import-to-supabase.mjs [--dry-run]
 */
import { createClient } from '@supabase/supabase-js';
import { readFileSync, existsSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(__dirname, '..');

function loadEnv() {
  const envPath = resolve(ROOT, '.env');
  if (!existsSync(envPath)) return;
  for (const line of readFileSync(envPath, 'utf8').split('\n')) {
    const t = line.trim();
    if (!t || t.startsWith('#')) continue;
    const i = t.indexOf('=');
    if (i < 0) continue;
    const k = t.slice(0, i).trim();
    const v = t.slice(i + 1).trim();
    if (!process.env[k]) process.env[k] = v;
  }
}

function restaurantRow(r, { includeMapsUrls = true } = {}) {
  const row = {
    id: r.id,
    name: r.name,
    country_code: r.country_code ?? 'ES',
    region_code: r.region_code,
    region_name: r.region_name,
    city: r.city,
    verification_status: r.verification_status ?? 'to_be_verified',
    is_published: r.is_published ?? true,
    is_hidden: r.is_hidden ?? false,
  };
  let optional = [
    'slug', 'province', 'district', 'postal_code', 'address_street',
    'latitude', 'longitude', 'venue_type', 'price_range',
    'last_verified_at', 'national_authority', 'phone', 'whatsapp',
    'email', 'website', 'menu_url', 'google_maps_url', 'apple_maps_url',
    'instagram', 'facebook', 'opening_hours',
    'seasonal_closure', 'description_es', 'description_en', 'description_de',
    'featured_image_url',
  ];
  if (!includeMapsUrls) {
    optional = optional.filter((k) => k !== 'google_maps_url' && k !== 'apple_maps_url');
  }
  for (const key of optional) {
    if (r[key] != null && r[key] !== '') row[key] = r[key];
  }
  for (const key of ['cuisine_types', 'meal_types', 'verification_methods']) {
    if (r[key]?.length) row[key] = r[key];
  }
  for (const key of ['face_program', 'aoecs_certified', 'is_premium_partner']) {
    if (r[key] != null) row[key] = r[key];
  }
  return row;
}

async function upsertLinks(client, restaurantId, links, table) {
  for (const link of links ?? []) {
    if (!link.platform) continue;
    const { error } = await client.from(table).upsert(
      {
        restaurant_id: restaurantId,
        platform: link.platform,
        url: link.url ?? '',
        is_active: link.is_active ?? true,
      },
      { onConflict: 'restaurant_id,platform' },
    );
    if (error) throw new Error(`${table} ${restaurantId}/${link.platform}: ${error.message}`);
  }
}

async function main() {
  const dryRun = process.argv.includes('--dry-run');
  loadEnv();

  const url = process.env.SUPABASE_URL;
  const key =
    process.env.SUPABASE_SERVICE_ROLE_KEY ?? process.env.SUPABASE_SECRET_KEY;
  if (!url || !key) {
    console.error('Fehler: SUPABASE_URL und SUPABASE_SERVICE_ROLE_KEY in .env');
    process.exit(1);
  }

  const jsonPath = resolve(ROOT, 'src/data/restaurants.json');
  const payload = JSON.parse(readFileSync(jsonPath, 'utf8'));
  const restaurants = payload.restaurants ?? [];
  console.log(`Lade ${restaurants.length} Restaurants aus restaurants.json`);

  if (dryRun) {
    console.log('Dry-run — keine Schreiboperationen.');
    return;
  }

  const client = createClient(url, key);

  let includeMapsUrls = true;
  const { error: schemaProbe } = await client
    .from('restaurants')
    .select('google_maps_url,apple_maps_url')
    .limit(0);
  if (schemaProbe) {
    includeMapsUrls = false;
    console.warn(
      'Hinweis: google_maps_url/apple_maps_url fehlen in Supabase — Import ohne Maps-URLs.',
    );
    console.warn('Bitte supabase/pending_maps_and_menu.sql ausführen, dann erneut importieren.');
  }

  for (const item of restaurants) {
    const rid = item.id;
    const { error: rErr } = await client
      .from('restaurants')
      .upsert(restaurantRow(item, { includeMapsUrls }), { onConflict: 'id' });
    if (rErr) throw new Error(`restaurants ${rid}: ${rErr.message}`);

    await upsertLinks(client, rid, item.delivery_links, 'delivery_links');
    await upsertLinks(client, rid, item.reservation_links, 'reservation_links');
  }

  console.log(`Fertig: ${restaurants.length} Restaurants in Supabase.`);
}

main().catch((e) => {
  console.error(e.message ?? e);
  process.exit(1);
});

File: scripts/import-to-supabase.py
#!/usr/bin/env python3
"""
Importiert src/data/restaurants.json nach Supabase (EU).

Voraussetzungen:
  pip install -r scripts/requirements.txt
  .env mit SUPABASE_URL und SUPABASE_SERVICE_ROLE_KEY

Verwendung:
  python scripts/import-to-supabase.py
  python scripts/import-to-supabase.py --dry-run
"""

from __future__ import annotations

import argparse
import json
import os
import sys
from pathlib import Path

try:
    from dotenv import load_dotenv
    from supabase import Client, create_client
except ImportError:
    print("Fehler: pip install python-dotenv supabase")
    sys.exit(1)

ROOT = Path(__file__).resolve().parent.parent
DEFAULT_JSON = ROOT / "src" / "data" / "restaurants.json"


def restaurant_row(r: dict) -> dict:
    """Flache Restaurant-Zeile für Postgres."""
    row = {
        "id": r["id"],
        "name": r["name"],
        "country_code": r.get("country_code", "ES"),
        "region_code": r["region_code"],
        "region_name": r["region_name"],
        "city": r["city"],
        "verification_status": r.get("verification_status", "to_be_verified"),
        "is_published": r.get("is_published", True),
        "is_hidden": r.get("is_hidden", False),
    }
    optional = [
        "slug", "province", "district", "postal_code", "address_street",
        "latitude", "longitude", "venue_type", "price_range",
        "last_verified_at", "national_authority", "phone", "whatsapp",
        "email", "website", "menu_url", "google_maps_url", "apple_maps_url",
        "instagram", "facebook", "opening_hours",
        "seasonal_closure", "description_es", "description_en", "description_de",
        "featured_image_url",
    ]
    for key in optional:
        if r.get(key) not in (None, ""):
            row[key] = r[key]
    for key in ("cuisine_types", "meal_types", "verification_methods"):
        if r.get(key):
            row[key] = r[key]
    for key in ("face_program", "aoecs_certified"):
        if r.get(key) is not None:
            row[key] = r[key]
    return row


def upsert_links(client: Client, restaurant_id: str, links: list, table: str) -> None:
    for link in links:
        if not link.get("platform"):
            continue
        client.table(table).upsert(
            {
                "restaurant_id": restaurant_id,
                "platform": link["platform"],
                "url": link.get("url") or "",
                "is_active": link.get("is_active", True),
            },
            on_conflict="restaurant_id,platform",
        ).execute()


def main() -> None:
    parser = argparse.ArgumentParser(description="JSON → Supabase importieren")
    parser.add_argument("--input", type=Path, default=DEFAULT_JSON)
    parser.add_argument("--dry-run", action="store_true")
    args = parser.parse_args()

    load_dotenv(ROOT / ".env")

    url = os.getenv("SUPABASE_URL")
    key = (
        os.getenv("SUPABASE_SERVICE_ROLE_KEY")
        or os.getenv("SUPABASE_SECRET_KEY")
    )
    if not url or not key:
        print(
            "Fehler: SUPABASE_URL und SUPABASE_SERVICE_ROLE_KEY (oder SUPABASE_SECRET_KEY) in .env setzen"
        )
        sys.exit(1)

    if not args.input.exists():
        print(f"Fehler: {args.input} nicht gefunden")
        sys.exit(1)

    payload = json.loads(args.input.read_text(encoding="utf-8"))
    restaurants = payload.get("restaurants", [])
    print(f"Lade {len(restaurants)} Restaurants aus {args.input.name}")

    if args.dry_run:
        print("Dry-run — keine Schreiboperationen.")
        return

    client = create_client(url, key)

    for item in restaurants:
        rid = item["id"]
        client.table("restaurants").upsert(restaurant_row(item), on_conflict="id").execute()
        upsert_links(client, rid, item.get("delivery_links") or [], "delivery_links")
        upsert_links(client, rid, item.get("reservation_links") or [], "reservation_links")

    print(f"Fertig: {len(restaurants)} Restaurants in Supabase.")


if __name__ == "__main__":
    main()
